Add support for model signature verification

stefanberger opened the pull request:
# Essential Elements of an Effective PR Description Checklist

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.


## Purpose
Add support for model signature verification on local models. To do this, replicate many of the command line options from the model signing library's model_signing tool into vllm so that many use cases (different types of signatures) are addressed.

For the local model signature verification check early on whether a directory with the given name is available and if signature verification is requested by the user by passing the --verification-method option.

Signature verification for models downloaded from the Huggingface hub is not addressed in this patch. For this I am currently waiting for another release of the model_signing library where the necessary symlink support is provided.

## Test Plan

To verify this patch you may want to try the following signed model:

git clone https://huggingface.co/ibm-granite/granite-4.0-tiny-preview

Verify and run it with the following command line:

vllm serve \
	--verification-method sigstore \
	--identity Granite.Preview@ibm.com \
	--identity-provider https://sigstore.verify.ibm.com/oauth2 \
	~/models/granite-4.0-tiny-preview/

## Test Result

The following line must appear in the `vllm serve`:
```
(APIServer pid=4545) INFO 07-30 16:49:48 [config.py:431] Signature verification succeeded
```

## (Optional) Documentation Update

stefanberger committed:
Add support for model signature verification on local models

Add support for model signature verification on local models. To do this,
replicate many of the command line options from the model signing library's
model_signing tool into vllm so that many use cases (different types of
signatures) are addressed.

For the local model signature verification check early on whether a
directory with the given name is available and if signature verification
is requested by the user by passing the --verification-method option.

Signature verification for models downloaded from the Huggingface hub
is not addressed in this patch.

To verify this patch you may want to try the following signed model:

git clone https://huggingface.co/ibm-granite/granite-4.0-tiny-preview

Verify and run it with the following command line:

vllm serve \
        --verification-method sigstore \
        --identity Granite.Preview@ibm.com \
        --identity-provider https://sigstore.verify.ibm.com/oauth2 \
        ~/models/granite-4.0-tiny-preview/

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger commented:
It looks like we are running in some sort of dependency issue with the patch for explicit installation of sigstore-protobuf-specs that I added so that this package gets installed at all (it was missing in previous builds).

```
2025-07-30T20:37:28.2218418Z #10 0.414 Using Python 3.12.11 environment at: /opt/venv
2025-07-30T20:37:29.5220108Z #10 1.864   × No solution found when resolving dependencies:
2025-07-30T20:37:29.5911288Z #10 1.865   ╰─▶ Because there is no version of betterproto==2.0.0b5 and
2025-07-30T20:37:29.5912023Z #10 1.865       sigstore-protobuf-specs<=0.2.1 depends on betterproto==2.0.0b5, we can
2025-07-30T20:37:29.5912699Z #10 1.865       conclude that sigstore-protobuf-specs<=0.2.1 cannot be used.
2025-07-30T20:37:29.5913477Z #10 1.865       And because only the following versions of sigstore-protobuf-specs are
2025-07-30T20:37:29.5914133Z #10 1.865       available:
2025-07-30T20:37:29.5914529Z #10 1.865           sigstore-protobuf-specs==0.1.0
2025-07-30T20:37:29.5915030Z #10 1.865           sigstore-protobuf-specs==0.2.0
2025-07-30T20:37:29.5915535Z #10 1.865           sigstore-protobuf-specs==0.2.1
2025-07-30T20:37:29.5916056Z #10 1.865           sigstore-protobuf-specs==0.2.2
2025-07-30T20:37:29.5916837Z #10 1.865           sigstore-protobuf-specs==0.3.0
2025-07-30T20:37:29.5917374Z #10 1.865           sigstore-protobuf-specs==0.3.1
2025-07-30T20:37:29.5917843Z #10 1.865           sigstore-protobuf-specs==0.3.2
2025-07-30T20:37:29.5918350Z #10 1.865           sigstore-protobuf-specs==0.3.3
2025-07-30T20:37:29.5918862Z #10 1.865           sigstore-protobuf-specs==0.3.4
2025-07-30T20:37:29.5919254Z #10 1.865           sigstore-protobuf-specs==0.3.5
2025-07-30T20:37:29.5919723Z #10 1.865           sigstore-protobuf-specs==0.4.0
2025-07-30T20:37:29.5920040Z #10 1.865           sigstore-protobuf-specs==0.4.1
2025-07-30T20:37:29.5920606Z #10 1.865           sigstore-protobuf-specs==0.4.2
2025-07-30T20:37:29.5920933Z #10 1.865           sigstore-protobuf-specs==0.4.3
2025-07-30T20:37:29.5921238Z #10 1.865           sigstore-protobuf-specs==0.5.0
2025-07-30T20:37:29.5921780Z #10 1.865       we can conclude that sigstore-protobuf-specs<0.2.2 cannot be used. (1)
2025-07-30T20:37:29.5922381Z #10 1.865 
2025-07-30T20:37:29.5922628Z #10 1.865       Because there is no version of betterproto==2.0.0b6 and
2025-07-30T20:37:29.5923047Z #10 1.865       sigstore-protobuf-specs>=0.2.2,<=0.3.2 depends on betterproto==2.0.0b6,
2025-07-30T20:37:29.5923496Z #10 1.865       we can conclude that sigstore-protobuf-specs>=0.2.2,<=0.3.2 cannot be
2025-07-30T20:37:29.5923818Z #10 1.865       used.
2025-07-30T20:37:29.5924107Z #10 1.865       And because we know from (1) that sigstore-protobuf-specs<0.2.2 cannot
2025-07-30T20:37:29.5924539Z #10 1.865       be used, we can conclude that sigstore-protobuf-specs<0.3.3 cannot be
2025-07-30T20:37:29.5924875Z #10 1.865       used.
2025-07-30T20:37:29.5925120Z #10 1.865       And because sigstore-protobuf-specs==0.3.3 was yanked
2025-07-30T20:37:29.5925481Z #10 1.865       (reason: Unintended API breakage), we can conclude that
2025-07-30T20:37:29.5925850Z #10 1.865       sigstore-protobuf-specs<0.3.4 cannot be used. (2)
2025-07-30T20:37:29.5926121Z #10 1.865 
2025-07-30T20:37:29.5926558Z #10 1.865       Because there is no version of betterproto==2.0.0b7 and
2025-07-30T20:37:29.5927038Z #10 1.865       sigstore-protobuf-specs>=0.3.4 depends on betterproto==2.0.0b7, we can
2025-07-30T20:37:29.5927491Z #10 1.865       conclude that sigstore-protobuf-specs>=0.3.4 cannot be used.
2025-07-30T20:37:29.5927912Z #10 1.865       And because we know from (2) that sigstore-protobuf-specs<0.3.4 cannot
2025-07-30T20:37:29.5928338Z #10 1.865       be used, we can conclude that all versions of sigstore-protobuf-specs
2025-07-30T20:37:29.5928672Z #10 1.865       cannot be used.
2025-07-30T20:37:29.5928984Z #10 1.865       And because you require sigstore-protobuf-specs, we can conclude that
2025-07-30T20:37:29.5929356Z #10 1.865       your requirements are unsatisfiable.
2025-07-30T20:37:29.5929594Z #10 1.865 
2025-07-30T20:37:29.5929871Z #10 1.865       hint: Pre-releases are available for `sigstore-protobuf-specs` in the
2025-07-30T20:37:29.5930322Z #10 1.865       requested range (e.g., 0.3.0rc0), but pre-releases weren't enabled (try:
2025-07-30T20:37:29.5930679Z #10 1.865       `--prerelease=allow`)
2025-07-30T20:37:29.5930894Z #10 1.865 
2025-07-30T20:37:29.5931149Z #10 1.865       hint: `betterproto` was requested with a pre-release marker (e.g.,
2025-07-30T20:37:29.5931570Z #10 1.865       betterproto==2.0.0b5), but pre-releases weren't enabled (try:
2025-07-30T20:37:29.5931882Z #10 1.865       `--prerelease=allow`)
2025-07-30T20:37:29.5932420Z #10 ERROR: process "/bin/sh -c uv pip install --upgrade pip &&     uv pip install -r requirements/cpu.txt" did not complete successfully: exit code: 1
2025-07-30T20:37:29.5932934Z ------
2025-07-30T20:37:29.5933859Z  > [base-common 5/5] RUN --mount=type=cache,target=/root/.cache/uv     --mount=type=bind,src=requirements/common.txt,target=requirements/common.txt     --mount=type=bind,src=requirements/cpu.txt,target=requirements/cpu.txt     uv pip install --upgrade pip &&     uv pip install -r requirements/cpu.txt:
2025-07-30T20:37:29.5935051Z 1.865       And because you require sigstore-protobuf-specs, we can conclude that
2025-07-30T20:37:29.5935419Z 1.865       your requirements are unsatisfiable.
2025-07-30T20:37:29.5935652Z 1.865 
2025-07-30T20:37:29.5935916Z 1.865       hint: Pre-releases are available for `sigstore-protobuf-specs` in the
2025-07-30T20:37:29.5936544Z 1.865       requested range (e.g., 0.3.0rc0), but pre-releases weren't enabled (try:
2025-07-30T20:37:29.5937100Z 1.865       `--prerelease=allow`)
2025-07-30T20:37:29.5937304Z 1.865 
2025-07-30T20:37:29.5937558Z 1.865       hint: `betterproto` was requested with a pre-release marker (e.g.,
2025-07-30T20:37:29.5938121Z 1.865       betterproto==2.0.0b5), but pre-releases weren't enabled (try:
2025-07-30T20:37:29.5938430Z 1.865       `--prerelease=allow`)
2025-07-30T20:37:29.5938628Z ------
```

mihaimaruseac commented:
Oh, that's the [`uv` issue](https://github.com/sigstore/protobuf-specs/issues/288). But there's also https://github.com/sigstore/sigstore-python/issues/1049 and after that lands we need to make `model_transparency` work in the new world and do a new release.

hmellor commented on vllm/engine/arg_utils.py:
```diff
@@ -851,6 +864,29 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
+            description=SignatureVerificationConfig.__doc__,
+        )
+        sv_group.add_argument("--verification-method",
+                              choices=[f.value for f in VerificationMethod],
```
If you changed `VerificationMethod` to be a `Literal` then this special case would not be necessary, `get_kwargs` would automatically populate `choices` with the contents of the `Literal`

stefanberger replied:
Done.

hmellor commented on vllm/config.py:
```diff
@@ -272,6 +275,163 @@ def is_init_field(cls: ConfigType, name: str) -> bool:
+    public_key: Optional[str] = None
+    """Public key required for signature verification when using the 'key'
+    method."""
+    _verification_done: bool = False
+    """Set to True once signature verification was done."""
```
Since this is considered state for this config, you could hint it as a [typing.ClassVar](https://docs.python.org/3/library/dataclasses.html#class-variables) so that it's not considered a `field` in the saame way as the other class members.

Might this be useful?

stefanberger replied:
> Might this be useful?

I don't know. I don't want it to be a class variable, but one per instance:

"When ClassVar is used, static type checkers will flag an error if an attempt is made to assign a value to that variable through an instance of the class. However, this restriction is not enforced at runtime by Python itself."

stefanberger replied:
Sounds good.

hmellor commented on vllm/engine/arg_utils.py:
```diff
@@ -451,6 +453,16 @@ class EngineArgs:
     kv_sharing_fast_prefill: bool = \
         CacheConfig.kv_sharing_fast_prefill
 
+    verification_method: Optional[str] = None
+    signature: Optional[str] = None
+    ignore_git_paths: bool = True
+    identity: str = ""
+    identity_provider: str = ""
+    use_staging: bool = False
+    certificate_chain: Optional[Iterable[str]] = None
+    log_fingerprints: bool = False
+    public_key: Optional[str] = None
```
Please get the defaults from `SignatureVerificationConfig` so that they will always be synchronised. e.g.:

```
    verification_method: Optional[str] = SignatureVerificationConfig.verification_method
    ...
```

hmellor commented on vllm/engine/arg_utils.py:
```diff
@@ -854,6 +866,28 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
                                 **vllm_kwargs["additional_config"])
 
+        # Signature verification arguments
+        sv_kwargs = get_kwargs(SignatureVerificationConfig)
+        sv_group = parser.add_argument_group(
+            title="SignatureVerificationConfig",
+            description=SignatureVerificationConfig.__doc__,
+        )
+        sv_group.add_argument("--verification-method",
+                              **sv_kwargs["verification_method"])
+        sv_group.add_argument("--signature", **sv_kwargs["signature"])
+        sv_group.add_argument("--ignore-paths", **sv_kwargs["ignore_paths"])
+        sv_group.add_argument("--ignore-git-paths",
+                              **sv_kwargs["ignore_git_paths"])
+        sv_group.add_argument("--use-staging", **sv_kwargs["use_staging"])
+        sv_group.add_argument("--identity", **sv_kwargs["identity"])
+        sv_group.add_argument("--identity-provider",
+                              **sv_kwargs["identity_provider"])
+        sv_group.add_argument("--certificate-chain",
+                              **sv_kwargs["certificate_chain"])
+        sv_group.add_argument("--log-fingerprints",
+                              **sv_kwargs["log_fingerprints"])
+        sv_group.add_argument("--public-key", **sv_kwargs["public_key"])
```
Perfect!

hmellor commented on vllm/config/__init__.py:
```diff
@@ -8,17 +8,20 @@
 import inspect
 import json
+import os
 import textwrap
 import uuid
 import warnings
-from collections.abc import Mapping
+from collections.abc import Iterable, Mapping
 from contextlib import contextmanager
 from dataclasses import MISSING, Field, field, fields, is_dataclass, replace
 from functools import cached_property, lru_cache
 from importlib.util import find_spec
+from pathlib import Path
 from typing import (TYPE_CHECKING, Any, Callable, ClassVar, Literal, Optional,
                     Protocol, TypeVar, Union, cast, get_args)
 
+import model_signing
```
These imports can be removed because they're not only used in `signature_verification.py`

stefanberger replied:
Relying too much on pre-commit. It doesn't remove them. Thanks.

stefanberger commented on vllm/config/signature_verification.py:
```diff
@@ -0,0 +1,204 @@
+ConfigT = TypeVar("ConfigT", bound=ConfigType)
+
+
+def config(cls: ConfigT) -> ConfigT:
+    """
+    A decorator that ensures all fields in a dataclass have default values
+    and that each field has a docstring.
+
+    If a `ConfigT` is used as a CLI argument itself, the default value provided
+    by `get_kwargs` will be the result parsing a JSON string as the kwargs
+    (i.e. `ConfigT(**json.loads(cli_arg))`). However, if a particular `ConfigT`
+    requires custom construction from CLI (i.e. `CompilationConfig`), it can
+    have a `from_cli` method, which will be called instead.
+
+    Config validation is performed by the tools/validate_config.py
+    script, which is invoked during the pre-commit checks.
+    """
+    return cls
```
Yes, this got factored-out recently. Thanks.

stefanberger committed:
logger: Add a logger for model_signing and adjust level when needed

Add a logger for the model_signing module and set its default log level
to WARNING. When the user wants to display the fingerprints of certificate
then temporarily set the logging level to INFO.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger committed:
Have betterproto==2.0.0b6 installed by explicitly listing it

Since uv does not install the prerelease version 2.0.0b6 of betterproto,
explicitly list among the common packages to install. Also list the
required version of sigstore so that this also gets installed.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
